:farmer: check_msgs_launch_test failing in kilted by a python package not importing site module

### Generated by Generative AI

_No response_

### Operating System:

Ubuntu Noble

### ROS version or commit hash:

Kilted

### RMW implementation (if applicable):

rmw_zenoh, cyclonedds

### RMW Configuration (if applicable):

_No response_

### Client library (if applicable):

rclcpp

### 'ros2 doctor --report' output

_No response_

### Steps to reproduce issue

1. go to any of the jobs listed below
2. Build a Job
3. See test regressions

### Expected behavior

To build without test regressions

### Actual behavior

Building with test regressions

### Additional information

### Reference Builds
- [Kci__nightly-zenoh_ubuntu_noble_amd64#109](http://build.ros2.org/job/Kci__nightly-zenoh_ubuntu_noble_amd64/109)
- [Kci__nightly-cyclonedds_ubuntu_noble_amd64#109](http://build.ros2.org/job/Kci__nightly-cyclonedds_ubuntu_noble_amd64/109)

### Failing Tests
- launch_testing_examples.launch_testing_examples.check_msgs_launch_test.launch_testing_examples.check_msgs_launch_test

### Log Output
```
35: OK
35: [INFO] [python3-14]: sending signal 'SIGINT' to process[python3-14]
35: [python3-14] Fatal Python error: init_import_site: Failed to import the site module
35: [python3-14] Python runtime state: initialized
35: [python3-14] Traceback (most recent call last):
35: [python3-14]   File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
35: [python3-14]   File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
35: [python3-14]   File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
35: [python3-14]   File "<frozen importlib._bootstrap>", line 1176, in exec_module
35: [python3-14]   File "<frozen site>", line 643, in <module>
35: [python3-14]   File "<frozen site>", line 630, in main
35: [python3-14]   File "<frozen site>", line 413, in addsitepackages
35: [python3-14]   File "<frozen site>", line 242, in addsitedir
35: [python3-14]   File "<frozen site>", line 201, in addpackage
35: [python3-14]   File "<string>", line 1, in <module>
35: [python3-14]   File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
35: [python3-14]   File "<frozen importlib._bootstrap>", line 1322, in _find_and_load_unlocked
35: [python3-14]   File "<frozen importlib._bootstrap>", line 1262, in _find_spec
35: [python3-14]   File "<frozen importlib._bootstrap_external>", line 1528, in find_spec
35: [python3-14]   File "<frozen importlib._bootstrap_external>", line 1502, in _get_spec
35: [python3-14]   File "<frozen importlib._bootstrap_external>", line 1605, in find_spec
35: [python3-14]   File "<frozen importlib._bootstrap_external>", line 1648, in _fill_cache
35: [python3-14] KeyboardInterrupt
35: [ERROR] [python3-14]: process has died [pid 292555, exit code 1, cmd '/usr/bin/python3 -c 
35: import signal
35: import time
35: 
35: try:
35:     while True:
35:         time.sleep(1)
35: except KeyboardInterrupt:
35:     pass
35: '].
35: test_process_terminates_as_expected[garbage, rmw_zenoh_cpp] (rcl.TestRMWImplementationIDCheckAfterShutdown.test_process_terminates_as_expected[garbage, rmw_zenoh_cpp])
35: Test that the executable under test terminates as expected. ... ok
```

### Flakyness 15 days
job_name|last_fail|first_fail|build_count|failure_count|failure_percentage
|--|--|--|--|--|--|
Kci__nightly-zenoh_ubuntu_noble_amd64|2026-03-31|2026-03-19|5|5|100.0
Kci__nightly-cyclonedds_ubuntu_noble_amd64|2026-03-31|2026-03-22|5|4|80.0
Rci__nightly-zenoh_ubuntu_noble_amd64|2026-03-30|2026-03-19|12|12|100.0

This error have been happenning for more than the jenkins history can save in memory, also even when it says 100% there are some builds that don't have this error.

CC: @Crola1702

--- a/rmw_test_fixture_implementation/src/rmw_test_fixture_implementation_py.c
+++ b/rmw_test_fixture_implementation/src/rmw_test_fixture_implementation_py.c
@@ -240,6 +240,9 @@ static rcutils_ret_t fini_array_and_contents(rcutils_array_list_t *array_list)
 static rmw_ret_t reload_environ(void)
 {
 #ifndef _WIN32
+  // Mask signals during environment reload to prevent fatal Python errors
+  // (init_import_site: Failed to import the site module) caused by signals
+  // interrupting sub-interpreter initialization.
   sigset_t set, oldset;
   sigemptyset(&set);
   sigaddset(&set, SIGINT);

--- a/rmw_test_fixture_implementation/src/rmw_test_fixture_implementation_py.c
+++ b/rmw_test_fixture_implementation/src/rmw_test_fixture_implementation_py.c
@@ -14,6 +14,10 @@
 
 #include <stdbool.h>
 
+#ifndef _WIN32
+#include <signal.h>
+#endif
+
 #include <Python.h>
 
 #include <rcutils/allocator.h>
@@ -235,10 +239,21 @@ static rcutils_ret_t fini_array_and_contents(rcutils_array_list_t *array_list)
  */
 static rmw_ret_t reload_environ(void)
 {
+#ifndef _WIN32
+  sigset_t set, oldset;
+  sigemptyset(&set);
+  sigaddset(&set, SIGINT);
+  sigaddset(&set, SIGTERM);
+  sigprocmask(SIG_BLOCK, &set, &oldset);
+#endif
+
   rcutils_allocator_t allocator = rcutils_get_default_allocator();
   rcutils_array_list_t pairs = rcutils_get_zero_initialized_array_list();
   rcutils_ret_t ret = rcutils_array_list_init(&pairs, 16, sizeof(rcutils_char_array_t), &allocator);
   if (ret) {
+#ifndef _WIN32
+    sigprocmask(SIG_SETMASK, &oldset, NULL);
+#endif
     return rmw_convert_rcutils_ret_to_rmw_ret(ret);
   }
 
@@ -248,12 +263,18 @@ static rmw_ret_t reload_environ(void)
   PyThreadState_Swap(orig_state);
   if (ret) {
     fini_array_and_contents(&pairs);
+#ifndef _WIN32
+    sigprocmask(SIG_SETMASK, &oldset, NULL);
+#endif
     return rmw_convert_rcutils_ret_to_rmw_ret(ret);
   }
 
   PyObject *os = PyImport_ImportModule("os");
   if (NULL == os) {
     fini_array_and_contents(&pairs);
+#ifndef _WIN32
+    sigprocmask(SIG_SETMASK, &oldset, NULL);
+#endif
     return RMW_RET_ERROR;
   }
 
@@ -261,6 +282,9 @@ static rmw_ret_t reload_environ(void)
   Py_DECREF(os);
   if (NULL == os_environ) {
     fini_array_and_contents(&pairs);
+#ifndef _WIN32
+    sigprocmask(SIG_SETMASK, &oldset, NULL);
+#endif
     return RMW_RET_ERROR;
   }
 
@@ -268,12 +292,19 @@ static rmw_ret_t reload_environ(void)
   if (NULL == res) {
     Py_DECREF(os_environ);
     fini_array_and_contents(&pairs);
+#ifndef _WIN32
+    sigprocmask(SIG_SETMASK, &oldset, NULL);
+#endif
     return RMW_RET_ERROR;
   }
 
   ret = pair_list_to_py_mapping(&pairs, os_environ);
   Py_DECREF(os_environ);
   fini_array_and_contents(&pairs);
+
+#ifndef _WIN32
+  sigprocmask(SIG_SETMASK, &oldset, NULL);
+#endif
 
   return rmw_convert_rcutils_ret_to_rmw_ret(ret);
 }
@@ -322,7 +353,7 @@ static PyMethodDef module_methods[] = {
 
 static struct PyModuleDef module = {
   PyModuleDef_HEAD_INIT,
-  "_rmw_test_fixtupre_implementation",
+  "_rmw_test_fixture_implementation",
   NULL,
   -1,
   module_methods,

